crates/sats/src/satn.rs

```diff
@@ -501,80 +510,171 @@ impl<'a, 'b> ser::SerializeNamedProduct for NamedFormatter<'a, 'b> {
     }
 }
 
-/// An implementation of [`Serializer`](ser::Serializer)
-/// that borrows from [`SatnFormatter`] except in `serialize_str`.
-struct PsqlFormatter<'a, 'b>(SatnFormatter<'a, 'b>);
+/// Provides the data format for named products for `SQL`.
+struct PsqlNamedFormatter<'a, 'b> {
+    /// The formatter for each element separating elements by a `,`.
+    f: EntryWrapper<'a, 'b, ','>,
+    /// The index of the element.
+    idx: usize,
+    /// If is not [Self::is_bytes_or_special] to control if we start with `(`
+    start: bool,
+    /// For checking [Self::is_bytes_or_special]
+    ty: &'a ProductType,
+    /// If the current element is a special type.
+    is_special: bool,
+}
+
+impl<'a, 'b> ser::SerializeNamedProduct for PsqlNamedFormatter<'a, 'b> {
+    type Ok = ();
+    type Error = SatnError;
+
+    fn serialize_element<T: Satn + ser::Serialize + ?Sized>(
+        &mut self,
+        name: Option<&str>,
+        elem: &T,
+    ) -> Result<(), Self::Error> {
+        // For binary data, output in `hex` format and skip the tagging of each value
+        self.is_special = ProductType::is_special_tag(name.unwrap_or_default());
+        self.f.entry(|mut f| {
+            if !self.is_special {
```

Comment:
Is it possible to have the special logic only for the binary case, and use the default formatter otherwise?

Reply:
Not because the `identity` and `Address` are product types and they cause the normal formatter to print the internal tag like`__identity_bytes`, that is the reason of all the roundabouts...